### DISCUSSION TITLE
How can I make a HTTP request in extensions?

### DISCUSSION THREAD

**Discussion by Kgg-moon** [microsoft/MicrosoftEdge-Extensions#83] [Q&A]
I designed an extensions that needs to make an http request, I didn't find the relevant API in the development documentation, so I used the following code, but it didn't work, is there a good brother to answer my confusion

**Comment by Kgg-moon** [comment]
code here
```
chrome.tabs.query({ currentWindow: true, active: true }, function (tabs) {
    var tab = tabs[0];
    chrome.scripting.executeScript({
        target: { tabId: tab.id },
        func: () => {
            return window.location.hostname;
        }
    }, (result) => {
        var domain = result[0].result;
        var main = getMainDomain(domain);


        document.getElementById('domain').textContent = domain;
        document.getElementById('main').textContent = main;
        var xhr = new XMLHttpRequest();
        xhr.open('GET', 'https://xxxxx + domain, true);
        xhr.onreadystatechange = function () {
            if (xhr.readyState === XMLHttpRequest.DONE && xhr.status === 200) {
                var response = xhr.responseText;
                console.log(response);
                document.getElementById('output').textContent = response;
            }
        };
        xhr.send();
    });
});
```

**Reply by sivsouvamMSFT** [reply]
Hi @Kgg-moon, thank you for reaching out to us. We will check with our Engineering team and revert soon.

**Comment by mukul-p** [comment]
Among other things, Extension on Chromium (i.e. Edge) has two main parts- [Extension Servicw Worker](https://developer.chrome.com/docs/extensions/develop/concepts/service-workers), and [Content Script](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts).

There are ways to connect with the web.

1. Extension Service Worker - There are two ways to connect to web from Service Worker
   - Use Fetch - https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers#replace-xmlhttprequest
   - Use WebSockets - https://developer.chrome.com/docs/extensions/how-to/web-platform/websockets

2. Content Script - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#xhr_and_fetch
